codalab/rest/json_api_client.py

```diff
@@ -0,0 +1,238 @@
+import base64
+from contextlib import closing
+import httplib
+import json
+import re
+import socket
+import sys
+import threading
+import time
+import urllib
+import urllib2
+import urlparse
+
+
+from codalab.client.bundle_client import BundleClient
+from codalab.common import http_error_to_exception
+
+
+def wrap_exception(message):
+    def decorator(f):
+        def wrapper(*args, **kwargs):
+            try:
+                return f(*args, **kwargs)
+            except JsonApiException as e:
+                raise JsonApiException, \
+                    JsonApiException(message + ': ' + e.message,
+                                           e.client_error), \
+                    sys.exc_info()[2]
+            except urllib2.HTTPError as e:
+                raise JsonApiException, \
+                    JsonApiException(message + ': ' +
+                                           httplib.responses[e.code] + ' - ' +
+                                           e.read(),
+                                           e.code >= 400 and e.code < 500), \
+                    sys.exc_info()[2]
+            except (urllib2.URLError, httplib.HTTPException, socket.error) as e:
+                raise JsonApiException, \
+                    JsonApiException(message + ': ' + str(e), False), \
+                    sys.exc_info()[2]
+        return wrapper
+    return decorator
+
+
+def unpack_json(raw_json):
```

Comment:
I experimented with deserializing on the client side with marshmallow, but I was struggling to think of a good way to represent, for example, the username of the owner on the client side. If we deserialize back into the original group_info, we get:

```
{
    "memberships": [3, 6, 7],
    "name": "ok-group",
    "owner_id": "6f2e70e0fed2487185fe0f21fffde98f",
    "uuid": "0x4612d33b6709410895b2bee028e05880"
}
```

We also get the owner object and each of the membership objects included as part of the compound response, but in order to look up the owner's username, we would have to search through these included objects to find the user matching the owner_id.

I instead opted to make the CLI/client side ignore any prior knowledge about the structure of the internal DB dicts, and instead just use these client-side JSON API attributes as is, and do a little bit of processing work to wire up the relationships. So instead we get this:

```
{
    "id": "0x4612d33b6709410895b2bee028e05880",
    "memberships": [
        {
            "group": {
                "id": "0x4612d33b6709410895b2bee028e05880"
            },
            "id": 3,
            "is_admin": true,
            "user": {
                "affiliation": "SCOr",
                "date_joined": "Mon Mar 28 22:49:17 2016",
                "email": "sckoo@cs.stanford.edu",
                "first_name": "\u9756\u5d34",
                "id": "6f2e70e0fed2487185fe0f21fffde98f",
                "last_name": "St\u00e9ph\u00e5n",
                "url": "http://sckoo.net/though",
                "user_name": "sckoo"
            }
        },
        {
            "group": {
                "id": "0x4612d33b6709410895b2bee028e05880"
            },
            "id": 6,
            "is_admin": true,
            "user": {
                "affiliation": null,
                "date_joined": "Fri Apr  8 20:53:24 2016",
                "email": "newuser@gmail.com",
                "first_name": null,
                "id": "fbeb0aebabb84da8b844ac9fba4fe95a",
                "last_name": null,
                "url": null,
                "user_name": "newuser"
            }
        },
        {
            "group": {
                "id": "0x4612d33b6709410895b2bee028e05880"
            },
            "id": 7,
            "is_admin": true,
            "user": {
                "affiliation": null,
                "date_joined": "Fri Apr  8 21:03:21 2016",
                "email": "newuser2@gmail.com",
                "first_name": null,
                "id": "9af28fa6a8e140e6a50561f3fcdb040e",
                "last_name": null,
                "url": null,
                "user_name": "newuser2"
            }
        }
    ],
    "name": "ok-group",
    "owner": {
        "affiliation": "SCOr",
        "date_joined": "Mon Mar 28 22:49:17 2016",
        "email": "sckoo@cs.stanford.edu",
        "first_name": "\u9756\u5d34",
        "id": "6f2e70e0fed2487185fe0f21fffde98f",
        "last_name": "St\u00e9ph\u00e5n",
        "url": "http://sckoo.net/though",
        "user_name": "sckoo"
    }
}
```

Now we can directly get the owner's username with `group['owner']['user_name']`. And we also don't need to choose the right deserializer for each of the objects in response, and we don't have to assume anything about how the objects are internally represented in the DB model. The `unpack_json` operation is the same for all responses.

Now I'm defining the function of marshmallow as: transforming db dicts into the JSON API format when formulating responses, and **validating** input data from JSON API requests and transforming them back into directly usable db dicts. I was able to completely factor out the Group ORMObject, and I'm confident I can simply the Bundle ORMObjects with this as well.